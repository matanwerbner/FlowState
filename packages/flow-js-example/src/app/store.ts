import { create } from 'zustand';
import {
  InitialData,
  InitialState,
  createStateMachine,
} from './lib/createStateMachine';
import { EventNames, GameContext, GameStates } from './lib/types';

type InitializeParams<GameContext> = {
  state?: GameStates;
  data?: Partial<GameContext>;
};

type Storetype = {
  currentState: GameStates;
  triggerEvent: (eventName: EventNames, data?: Partial<GameContext>) => void;
  context: GameContext;
  initializeStore: ({ state, data }: InitializeParams<GameContext>) => void;
};

export const useGameStore = create<Storetype>((set) => {
  const stateMachine = createStateMachine();

  return {
    context: stateMachine.context,
    currentState: stateMachine.currentState,
    triggerEvent: (eventName: EventNames, data?: Partial<GameContext>) => {
      return stateMachine.triggerEvent(eventName, data).then((newState) => {
        set({ currentState: newState, context: stateMachine.context });
      });
    },
    initializeStore: ({ state = InitialState, data = InitialData }) => {
      stateMachine.initialize(state, { ...InitialData, ...data });
      set({
        currentState: stateMachine.currentState,
        context: stateMachine.context,
      });
    },
  };
});
