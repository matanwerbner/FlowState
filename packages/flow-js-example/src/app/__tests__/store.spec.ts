// jest tests for store.ts

import { act } from '@testing-library/react';
import { useGameStore } from '../store';

describe('Store', () => {
  beforeEach(() => {
    useGameStore.getState().initializeStore({});
    jest.resetAllMocks();
  });

  it('should trigger pickCategory event and update state only if category is set', async () => {
    const { triggerEvent } = useGameStore.getState();
    expect(useGameStore.getState().currentState).toBe('pickCategory');
    await expect(triggerEvent('categoryPicked')).rejects.toMatchSnapshot();
    expect(useGameStore.getState().currentState).toBe('pickCategory');
    await act(
      async () => await triggerEvent('categoryPicked', { category: '9' })
    );
    expect(useGameStore.getState().currentState).toBe('pickDifficulty');
  });

  it('should trigger difficultyPicked event and update state only if category and difficulty are set', async () => {
    const { triggerEvent, initializeStore } = useGameStore.getState();
    initializeStore({ state: 'pickDifficulty', data: { category: '9' } });

    expect(useGameStore.getState().currentState).toBe('pickDifficulty');

    await act(
      async () =>
        await triggerEvent('difficultyPicked', {
          difficulty: 'easy',
          category: '9',
        })
    );
    expect(useGameStore.getState().currentState).toBe('fetchQuestions');
  });

  it("should trigger questionsFetched event and update state to 'gameInProgress'", async () => {
    const { initializeStore, triggerEvent } = useGameStore.getState();
    initializeStore({
      state: 'pickDifficulty',
      data: {
        category: '9',
      },
    });

    await act(
      async () =>
        await triggerEvent('difficultyPicked', {
          difficulty: 'easy',
        })
    );

    expect(useGameStore.getState().currentState).toBe('fetchQuestions');

    await expect(triggerEvent('questionsFetched')).rejects.toMatchSnapshot();

    await act(
      async () => await triggerEvent('questionsFetched', { questions: [] })
    );
    expect(useGameStore.getState().currentState).toBe('gameInProgress');
  });
});
