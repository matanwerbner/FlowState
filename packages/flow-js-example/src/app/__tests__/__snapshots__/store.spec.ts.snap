// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store should trigger pickCategory event and update state only if category is set 1`] = `
{
  "errors": {
    "category": "Category is required",
  },
}
`;

exports[`Store should trigger questionsFetched event and update state to 'gameInProgress' 1`] = `
{
  "errors": {
    "questions": "Questions not found",
  },
}
`;
